[Feature] Add grains to detect default gateway

basepi opened the pull request:
### What does this PR do?

Adds 3 grains, `ip_gw`, `ip4_gw`, and `ip6_gw` to the core grains, based on whether a default gateway shows up in `ip route show`.

### Tests written?

No

basepi committed:
Add grains to detect default gateway

salt-jenkins commented:
@basepi, thanks for your PR! By analyzing the history of the files in this pull request, we identified @terminalmage, @s0undt3ch and @thatch45 to be potential reviewers.

terminalmage commented:
Why not put the gateway IP in the grain? That could be useful.

basepi commented:
Mostly because this was easy, and all I needed for my hosts.

basepi commented:
I can probably do some logic massaging and testing tomorrow or Monday if we want the actual gateway IP.

terminalmage commented:
I personally think that having the gateway IP would be useful. @cachedout thoughts?

cachedout commented:
Yeah, I have to agree with @terminalmage on this one. As a new user, I wouldn't expect this to be a boolean. I'd expect it to contain the gateway itself. We can easily help modify this one to do that though. You know where to find us. :]

basepi commented:
Nah, it's easy, it will just take a little more testing which I won't have time to do until next week. I'll get it in.

damon-atkins commented:
Note you can have more than one default gateway, not that many people do.

Something that would be use full is the network the default gateway is on, which then can be used to determine all sorts of things, like picking the NTP and DNS server to use.

basepi committed:
Add first pass at ip parsing for default gateway

This is pretty naive parsing -- route configuration can be extensive,
can redirect to other tables, and is very hard to parse. We look for
`default via <ip>` and if we can't find it there we give up and just set
the grains to True.

basepi commented:
Right. The problem is that ip routing can be as complex as you want it to be. `default` can redirect to another routing table, there can be multiple `default` entries (though you'll only fall through if you don't get routed on those other entries), etc etc. I did a little bit of dumb parsing, just looking for the `default via` entry which appears to be the standard for simple routing tables. If I can't find it that way, I just set the grain to True instead.

If someone wants to try to make this smarter, more power to them. But beware migraines......

cachedout reviewed (changes requested):
I think that @damon-atkins has a reasonable point here but I can understand that's it's a rare case.

@basepi If you'll update the documentation to indicate that multiple gateways on the same host are not supported by this grain presently, then we can merge it, I think.

basepi committed:
Document lack of support for multiple default gateways

basepi commented:
@cachedout Done.
